README.md
# warp-server

[![MIT license][license-badge]][license-url]
[![Maintenance status][status-badge]][status-url]
[![Code coverage][coverage-badge]][coverage-url]

## About

**warp-server** is vanilla Node.js server for [warp] Google Chrome extension.

## Motivation

[warp] needs backend for at least one of its features. It's suitable time and place to gain a bit experience in vanilla Node. You can take a look at [list of questions and problems](./QA.md) that I faced during development.

## Requirements

Developed and tested on `Ubuntu 20.04.4 LTS`: `Node.js v12.22.12`.

## Usage

In fact this server could be used in any similar cases: push and pull JSON from remote server without authentication, data is not persisted if app failed.

Push data:

```
curl -X POST localhost:8081/somekey1 -i -H "Authorization: p@sw0rd" -d '{"number": 1234}'
```

Pull data:

```
curl -X GET localhost:8081/somekey1 -i
```

Update data:

```
curl -X PUT localhost:8081/somekey1 -i -H "Authorization: p@sw0rd" -d '{"number": 1234}'
```

Delete data:

```
curl -X DELETE localhost:8081/somekey1 -i -H "Authorization: p@sw0rd"
```

[status-url]: https://github.com/vikian050194/warp-server/pulse
[status-badge]: https://img.shields.io/github/last-commit/vikian050194/warp-server.svg

[license-url]: https://github.com/vikian050194/warp-server/blob/master/LICENSE
[license-badge]: https://img.shields.io/github/license/vikian050194/warp-server.svg

[coverage-url]: https://codecov.io/gh/vikian050194/warp-server
[coverage-badge]: https://img.shields.io/codecov/c/github/vikian050194/warp-server

[warp]: https://github.com/vikian050194/warp


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0]

### Added

- Push and pull data related to unique key
- Rewrite data

[Unreleased]: https://github.com/vikian050194/warp-server/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/vikian050194/warp-server/releases/tag/v0.1.0


tests/app-tests.js
const chai = require("chai");
const chaiHttp = require("chai-http");

const app = require("../src/app");

const should = chai.should();
chai.use(chaiHttp);

describe("app", function () {
    beforeEach(app.init);

    it("GET data - Not Found", (done) => {
        chai.request(app)
            .get("/test")
            .end((err, res) => {
                should.exist(res.status);
                res.should.have.status(404);
                done();
            });
    });

    it("POST data - OK", (done) => {
        const data = {
            "number": 1234
        };

        chai.request(app)
            .post("/test")
            .set("Authorization", "te$t")
            .send(data)
            .end((err, res) => {
                should.exist(res.status);
                res.should.have.status(200);
                done();
            });
    });

    it("PUT data - Not Found", (done) => {
        const data = {
            "number": 1234
        };

        chai.request(app)
            .put("/test")
            .set("Authorization", "te$t")
            .send(data)
            .end((err, res) => {
                should.exist(res.status);
                res.should.have.status(401);
                done();
            });
    });
});

src/app.js
const port = process.env.PORT || 8081;

const http = require("http");
const url = require("url");

const storage = new Map();
const secrets = new Map();

const init = () => {
    storage.clear();
    secrets.clear();
};

const server = http.createServer((req, res) => {
    let data = "";

    req.on("data", chunk => {
        data += chunk;
    });

    req.on("end", () => {
        const parsedUrl = url.parse(req.url, false);
        const urlChunks = parsedUrl.path.split("/").filter(item => item.length);
        const [team, ...tail] = urlChunks;
        const password = req.headers.authorization;

        if (tail.length) {
            res.writeHead(400);
            return;
        }

        switch (req.method) {
            case "GET": {
                if (storage.has(team) === false) {
                    res.writeHead(404);
                    break;
                }

                res.writeHead(200, { "Content-Type": "application/json" });
                res.write(JSON.stringify(storage.get(team)));
                break;
            }
            case "POST": {
                if (password === undefined) {
                    res.writeHead(401);
                    break;
                }

                if (storage.has(team) === true) {
                    res.writeHead(409);
                    break;
                }

                secrets.set(team, password);
                storage.set(team, JSON.parse(data));
                res.writeHead(200);
                break;
            }
            case "PUT": {
                if (password === undefined || secrets.get(team) !== password) {
                    res.writeHead(401);
                    break;
                }

                if (storage.has(team) === false) {
                    res.writeHead(404);
                    break;
                }

                storage.set(team, JSON.parse(data));
                res.writeHead(200);
                break;
            }
            case "DELETE": {
                if (password === undefined || secrets.get(team) !== password) {
                    res.writeHead(401);
                    break;
                }

                if (storage.has(team) === false) {
                    res.writeHead(404);
                    break;
                }

                secrets.delete(team);
                storage.delete(team);
                res.writeHead(200);
                break;
            }
        }

        res.end();
    });
}).listen(port, () => {
    if (process.env.NODE_ENV === "test") {
        console.info("Server is listening on port", port);
    }
});

module.exports = server;
module.exports.init = init;